eth2/utils/merkle_proof/src/lib.rs

```diff
@@ -73,6 +85,68 @@ impl MerkleTree {
         }
     }
 
+    /// Push an element in the MerkleTree.
+    /// MerkleTree and depth must be correct, as the algorithm expects valid data.
+    pub fn push_leaf(&mut self, elem: H256, depth: usize) -> Result<(), MerkleTreeError> {
+        use MerkleTree::*;
+
+        if depth == 0 {
+            return Err(MerkleTreeError::DepthTooSmall);
+        }
+
+        let mut right;
+        let mut left;
+        match &*self {
+            Leaf(_) => return Err(MerkleTreeError::LeafReached),
+            Zero(_) => *self = MerkleTree::create(&[elem], depth),
+            Node(_, l, r) => {
+                match (&**l, &**r) {
+                    // Tree is full
+                    (Leaf(_), Leaf(_)) => return Err(MerkleTreeError::MerkleTreeFull),
+                    // There is a right node so insert in right node
+                    (Node(_, _, _), Node(_, _, _)) => {
+                        right = *r.clone();
+                        if let Err(e) = right.push_leaf(elem, depth - 1) {
+                            return Err(e);
+                        }
+                        left = *l.clone();
```

Comment:
These clones are potentially quite heavy. I did some borrowck whispering to see if they could be removed, and it turns out they can if you juggle your references just right. My neatest attempt at a solution is here, but feel free to do something different:

https://github.com/michaelsproul/lighthouse/commit/2ca79f67174e9ac3cdf6550a52ff20e3ef68e0c1

Reply:
Added your commit to PR. Thanks for your help! Definitely learned a lot with this mem::replace :)